### PR TITLE
[MRG] Allow redirections in media files downloads

### LINKED ISSUE #2004: MediaPipeline (and ImagesPipeline/FilesPipeline) does not handle HTTP redirections
Basically, what's happened is that my spider is unable to download the files because the file_urls provided are actually redirected to the final download link. However, because of the following code, the redirect download middleware is effectively disabled, which makes the download fail.

```
def _check_media_to_download(self, result, request, info):
        if result is not None:
            return result
        if self.download_func:
            # this ugly code was left only to support tests. TODO: remove
            dfd = mustbe_deferred(self.download_func, request, info.spider)
            dfd.addCallbacks(
                callback=self.media_downloaded, callbackArgs=(request, info),
                errback=self.media_failed, errbackArgs=(request, info))
        else:
            request.meta['handle_httpstatus_all'] = True
            dfd = self.crawler.engine.download(request, info.spider)
            dfd.addCallbacks(
                callback=self.media_downloaded, callbackArgs=(request, info),
                errback=self.media_failed, errbackArgs=(request, info))
        return dfd
```

And here is the check in the redirect middleware that disables it:

```
if (request.meta.get('dont_redirect', False) or
                response.status in getattr(spider, 'handle_httpstatus_list', []) or
                response.status in request.meta.get('handle_httpstatus_list', []) or
                request.meta.get('handle_httpstatus_all', False)):
            return response

```

My question is: What is the point of enabling the handling of httpstatus_all, when it effectively disables all of the checks?

### PR DESCRIPTION
**Pull request by redapple**
Continuation of #2193
Fixes #2004 and #2449

Compared to #2193, I chose to remove `MEDIA_HTTPSTATUS_LIST`: I don't think this level of detail on status codes is needed for redirections. Debatable.

### CODE REVIEW

File: tests/test_pipeline_media.py [hunk 1]
```diff
@@ -249,3 +257,61 @@ def test_use_media_to_download_result(self):
         self.assertEqual(new_item['results'], [(True, 'ITSME')])
         self.assertEqual(self.pipe._mockcalled, \
                 ['get_media_requests', 'media_to_download', 'item_completed'])
+
+
+class MediaPipelineAllowRedirectSettingsTestCase(unittest.TestCase):
```
Comment (kmike) [inline tests/test_pipeline_media.py]: How hard would it be to have an integration-style test for this?

Comment (redapple) [inline tests/test_pipeline_media.py]: I'll have a go at it.

Comment (redapple) [inline tests/test_pipeline_media.py]: @kmike , I added some resources to MockServer to list files and redirect, and used similar tests to `tests/test_crawl.py`

**Comment by kmike** [comment]
Looks good, thanks @redapple!

**Comment by kmike** [comment]
@redapple how does it fix #2449?

**Comment by redapple** [comment]
@kmike , looks like I misread #2449. It has nothing to do with redirects.

**Comment by lopuhin** [comment]
@kmike @redapple what do you think of adding ``MEDIA_ALLOW_REDIRECTS = True`` to the new project template, so that it's on by default for new projects?

**Comment by redapple** [comment]
@lopuhin , sounds fine to me.

**Comment by lopuhin** [comment]
@redapple thanks, I'll make a PR!

**Comment by lopuhin** [comment]
Ah, although I see that media pipeline is not enabled in the sample project settings, so it might seem strange to have a setting that does nothing... And it's already documented, so someone who will be enabling this pipeline will see this setting too.

**Pull request merged**